src/hex_api_key.erl

```diff
@@ -100,10 +101,23 @@ get(Config, Name) when is_map(Config) and is_binary(Name) ->
 %% @end
 -spec add(hex_core:config(), binary(), [permission()]) -> hex_api:response().
 add(Config, Name, Permissions) when is_map(Config) and is_binary(Name) and is_list(Permissions) ->
+    assert_valid_permissions(Permissions),
     Path = hex_api:build_organization_path(Config, ["keys"]),
     Params = #{<<"name">> => Name, <<"permissions">> => Permissions},
     hex_api:post(Config, Path, Params).
 
+assert_valid_permissions(Permissions) ->
+    Pred = fun(Map) -> [assert_permission_key_value(K, V, Map) || {K,V} <- maps:to_list(Map)] end,
+    lists:foreach(Pred, Permissions).
+
+assert_permission_key_value(K, V, Permission) when not is_binary(K) orelse not is_binary(V) ->
+     Msg = "expected permissions to be a map with binary keys and values, got: ",
+     Err = iolist_to_binary([Msg, io_lib:format("~p", [Permission])]),
+     erlang:error({error, Err});
```

Comment:
could we return the tuple instead of raising the error? Sorry for confusion, I thought you did not want to raise the error :)

Reply:
No worries, I value the communication. I'd rather have some back and forth now vs reversing a decision later :) 

I think if we're going to return vs raise then it should not be in the form of `{error, binary()}` it should probably be `{error, {some_tag, binary}}` as the caller should know what do with the error easily. 

I'm glad you raised this (no pun intended) as I thought about this and raising doesn't feel right as it means as a consumer of hex_core I'll have to get defensive and wrap my calls to this module and so forth in `try` when really what I want from these calls is an ok tuple or error tuple, and if I get an error tuple I need to raise this in rebar3_hex as an example, which rebar3 will catch and pass back to a rebar3_hex provider:format_error/1 callback and that's precisely where I need to know what do to with the error. As it is and in the case of rebar3_hex, if it's just `{error, binary()` we will say "Unknown error, run with DEBUG=1 for detail" and that isn't very helpful. 

My final thought on this one is that a consuming library passing in an invalid set of k/vs in a permission map isn't exactly exceptional (but arguable). What do you think?

Reply:
> My final thought on this one is that a consuming library passing in an invalid set of k/vs in a permission map isn't exactly exceptional (but arguable)

yeah I can see it both ways too but I'm slightly leaning towards treating it as non-exceptional case and thus returning an error instead of raising it seems more appropriate.

> if it's just `{error, binary()}` we will say `"Unknown error, run with DEBUG=1 for detail"` and that isn't very helpful.

I think in this particular case the source of the error is rebar3_hex is using hex_core wrong, right? If we surface the error to the user, they can't still do anything about it, the user is not setting the types of things, rebar is. So I think `"Unknown error, run with DEBUG=1 for detail"` is actually what we want.

but in any case if `{error, {some_tag, binary}}` will help rebar in any way, I'm open to that. And just so it isn't lost, I earlier proposed `{error, string}` (string as in list) as that seems to be more conventional for some OTP functions. But no strong opinion either way.

what do you think?

Reply:
> > My final thought on this one is that a consuming library passing in an invalid set of k/vs in a permission map isn't exactly exceptional (but arguable)


> 
> but in any case if `{error, {some_tag, binary}}` will help rebar in any way, I'm open to that. And just so it isn't lost, I earlier proposed `{error, string}` (string as in list) as that seems to be more conventional for some OTP functions. But no strong opinion either way.
> 
> what do you think?

Well, I started this off with a list, but I think you didn't want to go that route because we end up with `[101,120,112,101,99,116,101,100,32,112,101,114,109,105,115,
 115,105,111,110,115,32,116,111,32,98,101,32,97,32|...]`  (https://github.com/hexpm/hex_core/pull/126#discussion_r656777539) and so forth. I don't mind either way whether it's a binary or string, if it's returned it vs raised I think it should be tagged though IMO. I would also be ok with ripping out the validation and just limiting this to type specs changes (which was a source of confusion to begin with) and just keep this limited to type spec and doc changes.

Reply:
@wojtekmach I suppose what I'll do is close this out, the code. It's really a tiny amount of code, and I think it's not clear what is desired. I would suggest making an issue or just copying / pasting what's in here. For now rebar3_hex will continue to suppress dialyzer warnings. Cheers.